### ISSUE TITLE
Merging/Modifying Types

### ISSUE THREAD

**Issue by ZeroIntensity** [python/typing#1490]
I have a decorator that modifies a class at runtime, kind of like a dataclass:

```py
@modify
class Something:
    ...
```

Let's say `modify` adds `foo` onto `Something`, so after calling `modify` you could use it like so:

```py
@modify
class Something
    ...

bar = Something.foo()
```

`dataclass_transform` doesn't support this kind of behavior at all. So, how would an API like this work?

I don't see why it would be difficult to implement some sort of `Merge` type, that takes in a type and "applies"" a `Protocol` to it. For the above example, it would work like this:

```py
T = TypeVar("T")

class FooProtocol(Protocol):
    @classmethod
    def foo() -> None:
        ...

def modify(tp: type[T]) -> Merge[T, FooProtocol]:
    tp.foo = ...  # type: ignore
    return tp

@modify
class Something:
    ...

bar = Something.foo()
```

In the above example, ideally it would just sort of mark that the `T` is now compatible with `FooProtocol`, but a `type: ignore` is used inside of `modify` because it would probably be a mess to try and come up with some sort of API for actually converting a `T` to `Merge[T, ...]`.

TL;DR A type that marks another type as being compatible with a protocol could be handy for making dataclass-like decorators.

**Comment by ZeroIntensity** [comment]
I have considered mixins, but it's not exactly ideal when you want to do more with the class than just modify it. For example, with a dataclass-like situation, I don't think the following would be very clear:

```py
class Foo(FooDataclass):
    a: str
    b: str
```

I guess you could resort to something like:

```py
@modify
class Foo(FooMixin):
    a: str
    b: str
```

But in my eyes that's too much boilerplate (especially if you're planning on doing it a lot). It is subjective, though. Is there a PEP on intersections yet?

**Comment by ViktorSky** [comment]
I have posted something similar, it would help during dynamic class overloads #1511
